### PR TITLE
Add Holt-Winters double exponential smoothing query function

### PR DESCRIPTION
**Pull request by eliothedeman**
This is the implementation of the topic discussed [here](https://groups.google.com/forum/#!topic/prometheus-developers/oQK0rdo7biQ)

Forecasting has not been implemented, but is relatively simple to add given the added helper functions.

### CODE REVIEW

**Comment by eliothedeman** [comment]
Here is an example 
screenshot from 2016-03-08 21-53-14

File: promql/functions.go [hunk 1]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+func funcHoltWinters(ev *evaluator, args Expressions) model.Value {
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// make an ouput vector large enough to hole the entire result
```
Comment (grobie) [inline promql/functions.go]: s/hole/hold

File: promql/bench.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package promql
```
Comment (fabxc) [inline promql/bench.go]: License header missing.

File: promql/functions_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package promql
```
Comment (fabxc) [inline promql/functions_test.go]: License header missing.

File: promql/functions.go [hunk 4]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// doubleSVal calculates the smoothed value at index i of raw data d
```
Comment (brian-brazil) [inline promql/functions.go]: This comment makes no sense unless you already understand what this function does. Can you add some context?

File: promql/functions.go [hunk 5]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+	y := (1 - tf) * doubleBVal(i-1, sf, tf, s, b, d)
+
+	// cache value
+	b[i] = x + y
+
+	return b[i]
+}
+
+// === holt_winters(node model.ValVector, smoothingFactor model.ValScalar, trendFactor model.ValScalar) Vector ===
+// algorhythm taken from https://en.wikipedia.org/wiki/Exponential_smoothing
```
Comment (brian-brazil) [inline promql/functions.go]: algorithm

Comment (brian-brazil) [inline promql/functions.go]: There's a number of algorithms listed on that page, can you indicate which one?

File: promql/functions.go [hunk 6]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// doubleSVal calculates the smoothed value at index i of raw data d
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
```
Comment (brian-brazil) [inline promql/functions.go]: Comments should start with a capital letter and end with a full stop.

File: promql/functions.go [hunk 7]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+	var s, b, d []float64
+
+	var l int
+	for _, samples := range mat {
+		l = len(samples.Values)
+
+		// can't do the smoothing operation with less than two points
+		if l > 2 {
+
+			// resize scratch values
```
Comment (brian-brazil) [inline promql/functions.go]: This will create new objects

File: promql/functions.go [hunk 8]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// make an ouput vector large enough to hold the entire result
+	resultVector := make(vector, len(mat.value()))
```
Comment (brian-brazil) [inline promql/functions.go]: I think you want `make(vector, 0, len(mat.value()))` and then you can remove the `resultCount` logic.

File: promql/bench.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// Test is a sequence of read and write commands that are run
+import "testing"
+
+// Benchmark runs a test multiple times as a built in benchmark
+type Benchmark struct {
+	b *testing.B
+	t *Test
+}
+
+// NewBenchmark returns an initilized empty Benchmark
```
Comment (mdlayher) [inline promql/bench.go]: "initialized", and add a period.

File: promql/functions.go [hunk 10]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// doubleSVal calculates the smoothed value at index i of raw data d
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
```
Comment (mdlayher) [inline promql/functions.go]: Remove empty line.  Looks weird to have blank space as the first line of a function body.

File: promql/functions.go [hunk 11]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+	// cache value
+	s[i] = x + y
+
+	return s[i]
+}
+
+// doubleBVal calculates the trend value at index i of raw data d
+// s = computed smoothed values, b = coumputed trend factors, d = raw
+func doubleBVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
```
Comment (mdlayher) [inline promql/functions.go]: Remove empty line.  Looks weird to have blank space as the first line of a function body.

File: promql/functions.go [hunk 12]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+	// create scratch values
+	var s, b, d []float64
+
+	var l int
+	for _, samples := range mat {
+		l = len(samples.Values)
+
+		// can't do the smoothing operation with less than two points
+		if l > 2 {
```
Comment (mdlayher) [inline promql/functions.go]: Change this to:

```go
if l < 2 {
    continue
}
```

Then, you can remove a level of indentation for the rest of the loop body.

File: promql/functions.go [hunk 13]
```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+
+			// resize scratch values
+			if l != len(s) {
+				s = make([]float64, l)
+				b = make([]float64, l)
+				d = make([]float64, l)
+			}
+
+			// NaN out stratch values, this use used to check for chached values
```
Comment (mdlayher) [inline promql/functions.go]: "cached"

File: promql/test.go [hunk 14]
```diff
@@ -411,6 +412,32 @@ func (cmd clearCmd) String() string {
 	return "clear"
 }
 
+// RunAsBenchmark runs the test in benchmarking mode.
+// This will not count any loads or non eval functions
+func (t *Test) RunAsBenchmark(b *testing.B) error {
+	for _, cmd := range t.cmds {
+
```
Comment (mdlayher) [inline promql/test.go]: Remove blank line.

File: promql/test.go [hunk 15]
```diff
@@ -411,6 +412,32 @@ func (cmd clearCmd) String() string {
...
+		switch cmd.(type) {
+
```
Comment (mdlayher) [inline promql/test.go]: Remove blank line.

File: promql/test.go [hunk 16]
```diff
@@ -411,6 +412,32 @@ func (cmd clearCmd) String() string {
...
+			b.StopTimer()
+			err := t.exec(cmd)
+			if err != nil {
+				return err
+			}
+			b.StartTimer()
+		}
+	}
+	return nil
+
```
Comment (mdlayher) [inline promql/test.go]: Remove blank line.

File: promql/functions_test.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    {path="/foo"} 20160
+`
+
+	bench := NewBenchmark(b, input)
+	bench.Run()
+}
+
+func BenchmarkDoubleSmooth1Day1Min(b *testing.B) {
+	input := `
+
```
Comment (mdlayher) [inline promql/functions_test.go]: Remove blank line.

**Comment by mdlayher** [comment]
Not sure what the Prometheus project typically does, but I'd recommend squashing into one commit with a message matching the PR title: "Add Holt-Winters double exponential smoothing query function".

There are too many small, insignificant commits, and merge commits right now, in my opinion.

File: promql/bench.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
+package promql
+
+// Test is a sequence of read and write commands that are run
```
Comment (mdlayher) [inline promql/bench.go]: Does the `import` statement really need a comment?

File: promql/bench.go [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+import "testing"
+
+// Benchmark runs a test multiple times as a built in benchmark
```
Comment (mdlayher) [inline promql/bench.go]: Add a period.

File: promql/bench.go [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+	if err != nil {
+		b.Fatalf("Unable to run benchmark: %s", err)
+	}
+	return &Benchmark{
+		b: b,
+		t: t,
+	}
+}
+
+// Run the benchmark a bunch of times
```
Comment (mdlayher) [inline promql/bench.go]: Add a period.

**Comment by grobie** [comment]
:+1: to squashing and that's what we usually do.

File: promql/functions.go (hunk 13 above)
Comment (mdlayher) [inline promql/functions.go]: "scratch"

Comment (mdlayher) [inline promql/functions.go]: s/use/is/ ?

File: promql/functions.go [hunk 21]
```diff
@@ -184,6 +184,132 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// sanity check input
+	if sf <= 0 || sf >= 1 {
+		ev.errorf("invalid trend factor. Expected: 0 < tf < 1 got: %f", sf)
```
Comment (mdlayher) [inline promql/functions.go]: Smoothing factor instead of trend factor?

**Pull request closed without merging**